sw/device/tests/pwrmgr_usb_clk_disabled_when_active_test.c

```diff
@@ -0,0 +1,120 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Tests the pwrmgr setting to disable the USB clock in active mode. The check
+// is to issue a USB CSR access when the clock is disabled, expecting the USB to
+// hung, and causing a watchdog reset.
+
+#include "sw/device/lib/dif/dif_aon_timer.h"
+#include "sw/device/lib/dif/dif_base.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rstmgr.h"
+#include "sw/device/lib/dif/dif_usbdev.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aon_timer_testutils.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/rstmgr_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "usbdev_regs.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static dif_aon_timer_t aon_timer;
+static dif_usbdev_t usbdev;
+
+static const uint32_t kExpectedHunkAddress =
+    TOP_EARLGREY_USBDEV_BASE_ADDR + USBDEV_INTR_ENABLE_REG_OFFSET;
+
+static void usbdev_csr_access(void) {
+  CHECK_DIF_OK(dif_usbdev_irq_set_enabled(&usbdev, kDifUsbdevIrqPowered,
+                                          kDifToggleEnabled));
+  dif_toggle_t state;
+  CHECK_DIF_OK(
+      dif_usbdev_irq_get_enabled(&usbdev, kDifUsbdevIrqPowered, &state));
+  CHECK(state == kDifToggleEnabled);
+}
+
+bool test_main(void) {
+  dif_pwrmgr_t pwrmgr;
+  dif_rstmgr_t rstmgr;
+
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+  CHECK_DIF_OK(dif_pwrmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+  CHECK_DIF_OK(dif_aon_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_AON_TIMER_AON_BASE_ADDR), &aon_timer));
+  CHECK_DIF_OK(dif_usbdev_init(
+      mmio_region_from_addr(TOP_EARLGREY_USBDEV_BASE_ADDR), &usbdev));
+
+  // Enable cpu dump capture.
+  CHECK_DIF_OK(dif_rstmgr_cpu_info_set_enabled(&rstmgr, kDifToggleEnabled));
+
+  if (UNWRAP(rstmgr_testutils_is_reset_info(&rstmgr, kDifRstmgrResetInfoPor))) {
```

Comment:
Just curious, what does `UNWRAP` macro do here?

Reply:
UNWRAP will log an error if the status_t returned indicates an error, otherwise it keeps going. See https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_framework/check.h;drc=f649352b711e0769653ad3baa5a3552f0a14cae9;l=288

Reply:
ah, thx.

Reply:
In the old times many testutils functions returned `void`, and internally logged fatal errors and aborted. This was changed to the current way where they return a status_t, and the caller decides what to do... like UNWRAP'ing it.